### PR TITLE
Fix: guard addPeer from adding peers who are closing

### PR DESCRIPTION
**Pull request by AlgoAxel**
On a closer read of the test failure [here](https://app.circleci.com/pipelines/github/algorand/go-algorand/12757/workflows/24ccc79c-2304-42aa-87ae-e28542c0f868/jobs/210769), it appears possible to add a peer to the wsNetwork which has already closed:

If the peer remotely closes, it will call the `wsNetwork.removePeer` function to remove itself. However, if this remote close happens before `addPeer` has added it, there is no net effect. Then, when `addPeer` is finally called, the peer is added without being checked that the peer is closed or not. A peer is added with a closed read/write loop, and it is not clear if/when the peer is ever discovered and cleaned up, besides connection performance monitoring.

This adds a simple guard clause from the closing channel on the peer to prevent adding peers which have already closed.

There is an opportunity for this closing channel to close *after* this guard check, but *before* addPeer does its work, so this does not eliminate this situation 100%. Since the peer and wsNetwork aren't synchronized to one another, we would have to put in some recurring checks (or save some state from `removePeer`) to see if any of the peers in the peers list have closed without us noticing.

### V2
now this guard uses the peer's `didSignalClose` atomic Int32. This flag gets set prior to the `removePeer` ever being called, so out of order calls should not have race potential.

Testing:
unit tests, still not reproducing locally (running now to attempt it more)

### CODE REVIEW

File: network/wsNetwork.go [hunk 1]
```diff
@@ -2455,6 +2455,13 @@ func (wn *WebsocketNetwork) removePeer(peer *wsPeer, reason disconnectReason) {
 }
 
 func (wn *WebsocketNetwork) addPeer(peer *wsPeer) {
+	// guard against peers which are already closing
+	select {
+	case <-peer.closing:
+		wn.log.Errorf("peer already closing %#v", peer)
```
Comment (algorandskiy) [inline network/wsNetwork.go:2461]: the fix looks right but I'm not sure if this should be an error

[the remark this answers is not in the document]
Comment (AlgoAxel) [reply network/wsNetwork.go:2461]: happy with that, sure

**Comment by excalq** [comment]
Telemetry gets a very high quantity of WARNING level entries having the message `read err: websocket: close 1006 (abnormal closure): unexpected EOF`. (About 5500/min). Would this guard against this condition, thereby reducing log noise?

**Review by algorandskiy — DISMISSED** [review]

**Comment by AlgoAxel** [comment]
@cce not sure how to directly "reply" to your comment,
> Is this mostly a test-only concern — I think my understanding is you are narrowing a race to make the test less flaky, but if this happened there is a periodic checker in the network that will notice the peer was closed, and remove it right?

I don’t think this is just a test concern. The guard I’ve implemented should prevent this out of order add, which will prevent this from happening, but to your point about “is there a periodic checker” — it appears not, but I could use your confirmation.
If the read or write loops close, then the closing channel is closed, and the didSignalClose flag(s) are also set, closing down all the peer’s threading.
However, when wsNetwork interacts with a peer, it does so by calling things like peer.writeNonBlockMsg() . this message gets dumped onto a prio/bulk outgoing channel which the loops would be processing, but they aren’t. Errors from trying to write to the closed connection would only ever happen again if the write loop was doing its job on the messages. Instead, writeNonBlockMsg returns a bool if it was enqueued or not, but even that boolean isn’t well used in the places its collected. So the write loop being closed isn’t currently well detected. And since the read loop is no longer running, it doesn’t seem like the read loop would cause wsNetwork to notice any errors at this point either.


https://github.dev/algorand/go-algorand/blob/3b197f48b185741c9e88cfa5f3df39fd4c123c93/network/wsPeer.go#L702-L703

**Review by algorandskiy — APPROVED** [review]

**Pull request merged**